Add configurable issuer/audience validation with dynamic claim support

diegocastrum opened the pull request:
This pull request introduces enhanced support for dynamic and static validation of JWT audience and issuer claims in Simple JWT, along with improved settings validation and more flexible claim handling. It adds new settings (`AUDIENCE_VALIDATION`, `ISSUER_CLAIM`), refines how audience and issuer are validated or set, and provides comprehensive documentation and tests for these behaviors.

**Settings and Validation Improvements:**

* Added the `AUDIENCE_VALIDATION` setting to control whether audience validation is "static" (default, enforced by PyJWT) or "dynamic" (enforced by Simple JWT logic after decoding). [[1]](diffhunk://#diff-d82241772fb86e472d94c9a04ee0802ed1d4cf03c9e688c0bbc723d78dceb077R23-R25) [[2]](diffhunk://#diff-1d78ba60e1efa4b7b08982e1bbaeceffefea079d402b7d58e57c0cfc7e6c78d8R27-R29) [[3]](diffhunk://#diff-1d78ba60e1efa4b7b08982e1bbaeceffefea079d402b7d58e57c0cfc7e6c78d8R160-R222)
* Added the `ISSUER_CLAIM` setting to allow customization of the issuer claim name (default: `"iss"`). [[1]](diffhunk://#diff-d82241772fb86e472d94c9a04ee0802ed1d4cf03c9e688c0bbc723d78dceb077R23-R25) [[2]](diffhunk://#diff-1d78ba60e1efa4b7b08982e1bbaeceffefea079d402b7d58e57c0cfc7e6c78d8R27-R29) [[3]](diffhunk://#diff-1d78ba60e1efa4b7b08982e1bbaeceffefea079d402b7d58e57c0cfc7e6c78d8R160-R222)
* Improved validation of settings: now checks that `AUDIENCE_VALIDATION` is a supported value and that `ISSUER` is a non-empty string or a sequence of non-empty strings, raising clear errors otherwise. [[1]](diffhunk://#diff-d82241772fb86e472d94c9a04ee0802ed1d4cf03c9e688c0bbc723d78dceb077R70-R144) [[2]](diffhunk://#diff-f20069a79c955c52519f37001105df187ba738b9a4dd49af010c723ff810200cR1-R39)

**Token Claim Handling:**

* Added methods to the `Token` class for setting and verifying audience (`aud`) and issuer (`iss`) claims, supporting both static and dynamic configurations, and handling both single and multiple allowed issuers/audiences. [[1]](diffhunk://#diff-e9863905dc4ef26b6bf2d8b6dc126dbb0e8ee574571c0c905af50933db5c1856R134-R139) [[2]](diffhunk://#diff-e9863905dc4ef26b6bf2d8b6dc126dbb0e8ee574571c0c905af50933db5c1856R152-R286) [[3]](diffhunk://#diff-e9863905dc4ef26b6bf2d8b6dc126dbb0e8ee574571c0c905af50933db5c1856R83-R84)
* Updated token creation and validation flows to ensure `iss` and `aud` claims are set and checked according to the new settings and logic. [[1]](diffhunk://#diff-e9863905dc4ef26b6bf2d8b6dc126dbb0e8ee574571c0c905af50933db5c1856R83-R84) [[2]](diffhunk://#diff-5192d507094cc6c3c4bbdc4280b4b25cd5d0db411335e208ad953d3511ecb59dR171)

**Backend and Decode Logic:**

* Updated the backend to allow skipping audience/issuer validation at the PyJWT decode step, deferring to Simple JWT logic when in dynamic mode or when multiple issuers are configured. [[1]](diffhunk://#diff-5789acf92aeca2ced1d02a664f5b7c4ca1db95e2269d0d4800167b79afe72d99R51-R52) [[2]](diffhunk://#diff-5789acf92aeca2ced1d02a664f5b7c4ca1db95e2269d0d4800167b79afe72d99R62-R69) [[3]](diffhunk://#diff-5789acf92aeca2ced1d02a664f5b7c4ca1db95e2269d0d4800167b79afe72d99L168-R180) [[4]](diffhunk://#diff-5fe5d26efdce206b257cb0c7c4568776b48bad5649082b83bb666af8bcb2022dL8-R9)

**Documentation and Testing:**

* Expanded documentation to explain the new settings and behaviors, including usage examples for dynamic audiences and issuers.
* Added and updated tests to cover the new settings validation, claim handling, and backend decode logic. [[1]](diffhunk://#diff-f20069a79c955c52519f37001105df187ba738b9a4dd49af010c723ff810200cR1-R39) [[2]](diffhunk://#diff-51d1cbb326cf0b05cde57945719d3b9c1f69f7d3f897ff247d19e0a0de7a3a38R299-R365) [[3]](diffhunk://#diff-09e1c5342323ae293ce8c6ee45eed0bfd1bf01a755d4d82dedbe9651e3a7f979R373-R411)

These changes make Simple JWT more flexible for multi-tenant or dynamic audience/issuer scenarios, and ensure that misconfigurations are detected early with clear error messages.

diegocastrum committed:
wip: add dynamic issuer support for JWT tokens

- Add `ISS_CLAIM` and `ALLOWED_ISSUERS` settings for issuer configuration
- Implement `set_iss()` and `verify_iss()` methods in `Token` class
- Add issuer verification in `Token.verify()` when configured
- Call `set_iss()` in `TokenRefreshSerializer` during token rotation
- Comment out issuer parameter from `TokenBackend` (WIP)
- Disable PyJWT's built-in issuer validation in `decode()`

This enables tokens to have dynamic issuers based on request context
while maintaining backward compatibility with static `ISSUER` setting.
Validation supports three modes:
- Static issuer (`ISSUER` setting)
- Whitelist validation (`ALLOWED_ISSUERS` setting)
- Dynamic issuers (no issuer validation beyond format checks)

Note: `TokenBackend` issuer handling still needs refactoring to fully
support the dynamic issuer workflow without double-decoding.

Fixes #899

diegocastrum committed:
fix: update issuer retrieval to use settings directly

diegocastrum committed:
feat: make issuer validation configurable and keep dynamic issuer support

  - add `ISSUER_VALIDATION` setting (static/dynamic) defaulting to static PyJWT enforcement
  - pass issuer to `TokenBackend/PyJWT` only in static mode; skip in dynamic mode to allow per-token issuers

Add support for `aud` (audience) and `iss` (issuer) claims in tokens
Fixes #899

diegocastrum committed:
feat: add configurable audience validation with dynamic mode

  - add `AUDIENCE_VALIDATION` setting (static/dynamic) and wire backend to skip PyJWT audience checks in dynamic mode
  - allow `Token.verify_aud` to accept per-token audiences when config is unset and dynamic mode is enabled
  - expose backend `verify_aud`/`verify_iss` flags and document the new setting

Add support for `aud` (audience) and `iss` (issuer) claims in tokens
Fixes #899

diegocastrum committed:
feat: add tests for configurable audience and issuer validation

diegocastrum committed:
feat: add support for dynamic audience and issuer claims in settings documentation

diegocastrum committed:
fix: add Optional import to typing for better type hinting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum committed:
docs: document issuer claim settings in SIMPLE_JWT

diegocastrum commented:
@Andrew-Chen-Wang would you mind taking a look when you have time? Would appreciate a quick review or guidance on how to move forward

Andrew-Chen-Wang commented on docs/settings.rst:
```diff
@@ -24,7 +24,11 @@ Some of Simple JWT's behavior can be customized through settings variables in
+      "AUDIENCE_VALIDATION": "static",
       "ISSUER": None,
+      "ISSUER_VALIDATION": "static",
+      "ISS_CLAIM": "iss",
```
Would be great to be consistent at least in the settings saying ISSUER_CLAIM. The code can say `iss` for short no prob.

diegocastrum replied:
Good call. I’ve renamed the public setting to `ISSUER_CLAIM` for consistency

Andrew-Chen-Wang commented on rest_framework_simplejwt/state.py:
```diff
@@ -5,8 +5,8 @@
-    api_settings.AUDIENCE,
-    api_settings.ISSUER,
+    api_settings.AUDIENCE if api_settings.AUDIENCE_VALIDATION == "static" else None,
+    api_settings.ISSUER if api_settings.ISSUER_VALIDATION == "static" else None,
```
is this none because ALLOWED_ISSUERS is there?

diegocastrum replied:
Yes, exactly. In dynamic mode I pass `issuer=None` so PyJWT does not enforce a single static issuer during decode. The issuer is then validated in `Token.verify_iss()`, with `ALLOWED_ISSUERS` acting as the optional whitelist.

Andrew-Chen-Wang commented on rest_framework_simplejwt/settings.py:
```diff
@@ -20,7 +20,11 @@
     "ISSUER": None,
+    "ISSUER_VALIDATION": "static",
+    "ISS_CLAIM": "iss",
+    "ALLOWED_ISSUERS": None,
```
Can you modify `__check_user_settings` in this file to make sure the user's settings look correct? I feel like a lot of settings combinations can be incorrect here.

diegocastrum replied:
Agreed. I tightened `__check_user_settings()` so invalid combinations fail early. In particular `ALLOWED_ISSUERS` now requires dynamic issuer validation, and I also reject the ambiguous `ISSUER_VALIDATION="dynamic"` + `ISSUER` + `ALLOWED_ISSUERS` combination because that mixes the static and dynamic issuer models

I added the validation checks and split them into small helpers in `settings.py` so the configuration rules stay centralized without turning `__check_user_settings()` into one large conditional block.

Andrew-Chen-Wang commented on docs/settings.rst:
```diff
@@ -24,7 +24,11 @@ Some of Simple JWT's behavior can be customized through settings variables in
       "ISSUER": None,
+      "ISSUER_VALIDATION": "static",
+      "ISS_CLAIM": "iss",
+      "ALLOWED_ISSUERS": None,
```
I feel like ISSUER could just be a list instead. If it's a list, we'll just automatically omit the iss key because the RFC doesn't allow lists in the first place (some implementations use it, but let's not since we could break some external/third-party apps)

diegocastrum replied:
I see the appeal of collapsing this into `ISSUER`, I was trying to keep the existing single string `ISSUER` behavior intact for token generation and static verification, and use `ALLOWED_ISSUERS` only for the dynamic verification. I will iterate on this one more time

diegocastrum replied:
@Andrew-Chen-Wang I reworked this to use `ISSUER` alone instead of introducing `ALLOWED_ISSUERS` and `ISSUER_VALIDATION`. A string keeps the existing static behavior, a list/tuple is treated as a verification-only issuer whitelist and iss is not emitted automatically in that case. I also added focused settings/token tests around the new shape.

Let me know what do you think about this

Andrew-Chen-Wang reviewed (changes requested):
thanks for these changes. left some comments. have more to come pending your thoughts

diegocastrum committed:
refactor: rename ISS_CLAIM to ISSUER_CLAIM for consistency

ref: https://github.com/jazzband/djangorestframework-simplejwt/pull/957#discussion_r3107797245

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum committed:
feat: add validation for issuer settings and implement corresponding tests

ref: https://github.com/jazzband/djangorestframework-simplejwt/pull/957#discussion_r3107806583

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum committed:
feat: support setting issuer as a string or sequence, update validation logic and tests

Fixes #899

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

diegocastrum commented:
@Andrew-Chen-Wang I’ve addressed the previous feedback and updated the PR accordingly. Would you mind taking another look when you have some time?

Thanks a lot!
